File: Core/Models/Tournament.swift
import Foundation

public enum WeightCategory: String, Codable, CaseIterable, Sendable, Hashable {
    case cubsUnder28
    case cubsUnder32
    case cubsUnder36
    case cubsOver36
    case cadetsUnder37
    case cadetsUnder45
    case cadetsUnder53
    case cadetsUnder61
    case cadetsOver61
    case juniorsUnder48
    case juniorsUnder55
    case juniorsUnder63
    case juniorsUnder73
    case juniorsOver73
    case seniorsUnder58
    case seniorsUnder68
    case seniorsUnder80
    case seniorsHeavy

    public var range: (lower: Double?, upper: Double?) {
        switch self {
        case .cubsUnder28: (nil, 28)
        case .cubsUnder32: (28, 32)
        case .cubsUnder36: (32, 36)
        case .cubsOver36: (36, nil)
        case .cadetsUnder37: (nil, 37)
        case .cadetsUnder45: (37, 45)
        case .cadetsUnder53: (45, 53)
        case .cadetsUnder61: (53, 61)
        case .cadetsOver61: (61, nil)
        case .juniorsUnder48: (nil, 48)
        case .juniorsUnder55: (48, 55)
        case .juniorsUnder63: (55, 63)
        case .juniorsUnder73: (63, 73)
        case .juniorsOver73: (73, nil)
        case .seniorsUnder58: (nil, 58)
        case .seniorsUnder68: (58, 68)
        case .seniorsUnder80: (68, 80)
        case .seniorsHeavy: (80, nil)
        }
    }

    public var ageGroup: AgeGroup {
        switch self {
        case .cubsUnder28, .cubsUnder32, .cubsUnder36, .cubsOver36: .cubs
        case .cadetsUnder37, .cadetsUnder45, .cadetsUnder53, .cadetsUnder61, .cadetsOver61: .cadets
        case .juniorsUnder48, .juniorsUnder55, .juniorsUnder63, .juniorsUnder73, .juniorsOver73: .juniors
        case .seniorsUnder58, .seniorsUnder68, .seniorsUnder80, .seniorsHeavy: .seniors
        }
    }

    public var labelKey: String { "weight.\(rawValue)" }

    public var shortLabel: String {
        let lower = range.lower
        let upper = range.upper
        if let upper, lower == nil { return "−\(Int(upper))kg" }
        if let lower, upper == nil { return "+\(Int(lower))kg" }
        if let lower, let upper { return "\(Int(lower))–\(Int(upper))kg" }
        return rawValue
    }

    /// Closest matching category given an athlete's age group + weight (treats
    /// `.kids` as cubs since WT competitions only start at cadet age).
    public static func suggested(for athlete: Athlete) -> WeightCategory? {
        let groupForCompetition: AgeGroup = athlete.ageGroup == .kids ? .cubs : athlete.ageGroup
        let weight = athlete.weightKg
        return Self.allCases.first { wc in
            guard wc.ageGroup == groupForCompetition else { return false }
            let lowerOK = wc.range.lower.map { weight > $0 } ?? true
            let upperOK = wc.range.upper.map { weight <= $0 } ?? true
            return lowerOK && upperOK
        }
    }
}

public enum HostingFederation: String, Codable, CaseIterable, Sendable, Hashable {
    case wtf, gcc, uae, clubInternal

    public var labelKey: String { "federation.\(rawValue)" }
}

/// Geographic / political reach of an event (Pillar 7).
public enum EventLevel: String, Codable, CaseIterable, Sendable, Hashable {
    case local, national, regional, international

    public var labelKey: String { "event_level.\(rawValue)" }

    /// Color used for level pills in the UI — escalates with reach.
    public var rank: Int {
        switch self {
        case .local: 1
        case .national: 2
        case .regional: 3
        case .international: 4
        }
    }
}

public struct Tournament: Codable, Identifiable, Hashable, Sendable {
    public let id: EntityID
    public var name: String
    public var nameAr: String?
    public var hostingFederation: HostingFederation
    public var startsAt: Date
    public var endsAt: Date
    public var location: String
    public var locationAr: String?
    public var isOfficial: Bool
    public var weightCategoriesOffered: [WeightCategory]

    /// Reach — local / national / regional / international (Pillar 7).
    public var level: EventLevel?
    /// Free-form name of the body sanctioning the event when more granular
    /// than `hostingFederation` (e.g. "UAE TKD Federation", "WT", "Asian TKD").
    public var sanctioningBody: String?

    public init(
        id: EntityID = UUID(),
        name: String,
        nameAr: String? = nil,
        hostingFederation: HostingFederation,
        startsAt: Date,
        endsAt: Date,
        location: String,
        locationAr: String? = nil,
        isOfficial: Bool,
        weightCategoriesOffered: [WeightCategory],
        level: EventLevel? = nil,
        sanctioningBody: String? = nil
    ) {
        self.id = id
        self.name = name
        self.nameAr = nameAr
        self.hostingFederation = hostingFederation
        self.startsAt = startsAt
        self.endsAt = endsAt
        self.location = location
        self.locationAr = locationAr
        self.isOfficial = isOfficial
        self.weightCategoriesOffered = weightCategoriesOffered
        self.level = level
        self.sanctioningBody = sanctioningBody
    }
}

public enum RegistrationStatus: String, Codable, CaseIterable, Sendable, Hashable {
    case registered, weighedIn, withdrawn, disqualified

    public var labelKey: String { "registration.\(rawValue)" }
}

public struct TournamentRegistration: Codable, Identifiable, Hashable, Sendable {
    public let id: EntityID
    public var tournamentID: EntityID
    public var athleteID: EntityID
    public var weightCategory: WeightCategory
    public var seedRank: Int?
    public var registeredAt: Date
    public var status: RegistrationStatus

    // === Pillar 7: per-event result ===
    public var ageDivisionEntered: AgeGroup?
    public var bracketSize: Int?
    public var finalPosition: Int?
    public var medal: MedalType?

    public init(
        id: EntityID = UUID(),
        tournamentID: EntityID,
        athleteID: EntityID,
        weightCategory: WeightCategory,
        seedRank: Int? = nil,
        registeredAt: Date = Date(),
        status: RegistrationStatus = .registered,
        ageDivisionEntered: AgeGroup? = nil,
        bracketSize: Int? = nil,
        finalPosition: Int? = nil,
        medal: MedalType? = nil
    ) {
        self.id = id
        self.tournamentID = tournamentID
        self.athleteID = athleteID
        self.weightCategory = weightCategory
        self.seedRank = seedRank
        self.registeredAt = registeredAt
        self.status = status
        self.ageDivisionEntered = ageDivisionEntered
        self.bracketSize = bracketSize
        self.finalPosition = finalPosition
        self.medal = medal
    }
}

public struct WeightCutEntry: Codable, Identifiable, Hashable, Sendable {
    public let id: EntityID
    public var registrationID: EntityID
    public var recordedAt: Date
    public var currentKg: Double
    public var targetKg: Double
    public var notes: String?

    public init(
        id: EntityID = UUID(),
        registrationID: EntityID,
        recordedAt: Date,
        currentKg: Double,
        targetKg: Double,
        notes: String? = nil
    ) {
        self.id = id
        self.registrationID = registrationID
        self.recordedAt = recordedAt
        self.currentKg = currentKg
        self.targetKg = targetKg
        self.notes = notes
    }

    public var deltaKg: Double { currentKg - targetKg }

    public func daysToCompetition(_ competitionDate: Date) -> Int {
        let cal = Calendar.current
        return cal.dateComponents([.day], from: recordedAt, to: competitionDate).day ?? 0
    }
}

public struct Bracket: Codable, Identifiable, Hashable, Sendable {
    public let id: EntityID
    public var tournamentID: EntityID
    public var weightCategory: WeightCategory
    public var seeds: [EntityID]
    public var generatedAt: Date

    public init(
        id: EntityID = UUID(),
        tournamentID: EntityID,
        weightCategory: WeightCategory,
        seeds: [EntityID],
        generatedAt: Date = Date()
    ) {
        self.id = id
        self.tournamentID = tournamentID
        self.weightCategory = weightCategory
        self.seeds = seeds
        self.generatedAt = generatedAt
    }
}

public struct BracketMatch: Codable, Identifiable, Hashable, Sendable {
    public let id: EntityID
    public var bracketID: EntityID
    public var round: Int
    public var position: Int
    public var athleteAID: EntityID?
    public var athleteBID: EntityID?
    public var winnerID: EntityID?
    public var matchID: EntityID?

    public init(
        id: EntityID = UUID(),
        bracketID: EntityID,
        round: Int,
        position: Int,
        athleteAID: EntityID? = nil,
        athleteBID: EntityID? = nil,
        winnerID: EntityID? = nil,
        matchID: EntityID? = nil
    ) {
        self.id = id
        self.bracketID = bracketID
        self.round = round
        self.position = position
        self.athleteAID = athleteAID
        self.athleteBID = athleteBID
        self.winnerID = winnerID
        self.matchID = matchID
    }
}

File: flutter_app/lib/core/models/tournament.dart
import 'athlete.dart';
import 'entity_id.dart';
import 'match.dart';

/// Port of Core/Models/Tournament.swift.

enum WeightCategory {
  cubsUnder28,
  cubsUnder32,
  cubsUnder36,
  cubsOver36,
  cadetsUnder37,
  cadetsUnder45,
  cadetsUnder53,
  cadetsUnder61,
  cadetsOver61,
  juniorsUnder48,
  juniorsUnder55,
  juniorsUnder63,
  juniorsUnder73,
  juniorsOver73,
  seniorsUnder58,
  seniorsUnder68,
  seniorsUnder80,
  seniorsHeavy;

  /// (lower, upper) — null means unbounded on that side.
  (double? lower, double? upper) get range => switch (this) {
    WeightCategory.cubsUnder28 => (null, 28),
    WeightCategory.cubsUnder32 => (28, 32),
    WeightCategory.cubsUnder36 => (32, 36),
    WeightCategory.cubsOver36 => (36, null),
    WeightCategory.cadetsUnder37 => (null, 37),
    WeightCategory.cadetsUnder45 => (37, 45),
    WeightCategory.cadetsUnder53 => (45, 53),
    WeightCategory.cadetsUnder61 => (53, 61),
    WeightCategory.cadetsOver61 => (61, null),
    WeightCategory.juniorsUnder48 => (null, 48),
    WeightCategory.juniorsUnder55 => (48, 55),
    WeightCategory.juniorsUnder63 => (55, 63),
    WeightCategory.juniorsUnder73 => (63, 73),
    WeightCategory.juniorsOver73 => (73, null),
    WeightCategory.seniorsUnder58 => (null, 58),
    WeightCategory.seniorsUnder68 => (58, 68),
    WeightCategory.seniorsUnder80 => (68, 80),
    WeightCategory.seniorsHeavy => (80, null),
  };

  AgeGroup get ageGroup => switch (this) {
    WeightCategory.cubsUnder28 ||
    WeightCategory.cubsUnder32 ||
    WeightCategory.cubsUnder36 ||
    WeightCategory.cubsOver36 => AgeGroup.cubs,
    WeightCategory.cadetsUnder37 ||
    WeightCategory.cadetsUnder45 ||
    WeightCategory.cadetsUnder53 ||
    WeightCategory.cadetsUnder61 ||
    WeightCategory.cadetsOver61 => AgeGroup.cadets,
    WeightCategory.juniorsUnder48 ||
    WeightCategory.juniorsUnder55 ||
    WeightCategory.juniorsUnder63 ||
    WeightCategory.juniorsUnder73 ||
    WeightCategory.juniorsOver73 => AgeGroup.juniors,
    WeightCategory.seniorsUnder58 ||
    WeightCategory.seniorsUnder68 ||
    WeightCategory.seniorsUnder80 ||
    WeightCategory.seniorsHeavy => AgeGroup.seniors,
  };

  String get labelKey => 'weight.$name';

  String get shortLabel {
    final r = range;
    final lower = r.$1;
    final upper = r.$2;
    if (upper != null && lower == null) return '−${upper.toInt()}kg';
    if (lower != null && upper == null) return '+${lower.toInt()}kg';
    if (lower != null && upper != null) {
      return '${lower.toInt()}–${upper.toInt()}kg';
    }
    return name;
  }

  /// Closest matching category given an athlete's age group + weight.
  static WeightCategory? suggested(Athlete athlete) {
    final groupForCompetition = athlete.ageGroup == AgeGroup.kids
        ? AgeGroup.cubs
        : athlete.ageGroup;
    final weight = athlete.weightKg;
    for (final wc in WeightCategory.values) {
      if (wc.ageGroup != groupForCompetition) continue;
      final r = wc.range;
      final lowerOk = r.$1 == null || weight > r.$1!;
      final upperOk = r.$2 == null || weight <= r.$2!;
      if (lowerOk && upperOk) return wc;
    }
    return null;
  }

  static WeightCategory fromJson(String raw) =>
      WeightCategory.values.firstWhere(
        (e) => e.name == raw,
        orElse: () => WeightCategory.seniorsHeavy,
      );
}

enum HostingFederation {
  wtf,
  gcc,
  uae,
  clubInternal;

  String get labelKey => 'federation.$name';

  static HostingFederation fromJson(String raw) =>
      HostingFederation.values.firstWhere(
        (e) => e.name == raw,
        orElse: () => HostingFederation.clubInternal,
      );
}

/// Geographic / political reach of an event.
enum EventLevel {
  local,
  national,
  regional,
  international;

  String get labelKey => 'event_level.$name';

  /// Escalates with reach; used for level pills in the UI.
  int get rank => switch (this) {
    EventLevel.local => 1,
    EventLevel.national => 2,
    EventLevel.regional => 3,
    EventLevel.international => 4,
  };

  static EventLevel fromJson(String raw) => EventLevel.values.firstWhere(
    (e) => e.name == raw,
    orElse: () => EventLevel.local,
  );
}

class Tournament {
  final EntityID id;
  final String name;
  final String? nameAr;
  final HostingFederation hostingFederation;
  final DateTime startsAt;
  final DateTime endsAt;
  final String location;
  final String? locationAr;
  final bool isOfficial;
  final List<WeightCategory> weightCategoriesOffered;

  /// Reach — local / national / regional / international.
  final EventLevel? level;

  /// Free-form name of the sanctioning body when more granular than
  /// [hostingFederation].
  final String? sanctioningBody;

  const Tournament({
    required this.id,
    required this.name,
    this.nameAr,
    required this.hostingFederation,
    required this.startsAt,
    required this.endsAt,
    required this.location,
    this.locationAr,
    required this.isOfficial,
    this.weightCategoriesOffered = const [],
    this.level,
    this.sanctioningBody,
  });

  factory Tournament.create({
    EntityID? id,
    required String name,
    String? nameAr,
    required HostingFederation hostingFederation,
    required DateTime startsAt,
    required DateTime endsAt,
    required String location,
    String? locationAr,
    required bool isOfficial,
    List<WeightCategory> weightCategoriesOffered = const [],
    EventLevel? level,
    String? sanctioningBody,
  }) => Tournament(
    id: id ?? newEntityId(),
    name: name,
    nameAr: nameAr,
    hostingFederation: hostingFederation,
    startsAt: startsAt,
    endsAt: endsAt,
    location: location,
    locationAr: locationAr,
    isOfficial: isOfficial,
    weightCategoriesOffered: weightCategoriesOffered,
    level: level,
    sanctioningBody: sanctioningBody,
  );

  factory Tournament.fromJson(Map<String, dynamic> json) => Tournament(
    id: json['id'] as String,
    name: json['name'] as String,
    nameAr: json['nameAr'] as String?,
    hostingFederation: HostingFederation.fromJson(
      json['hostingFederation'] as String,
    ),
    startsAt: DateTime.parse(json['startsAt'] as String),
    endsAt: DateTime.parse(json['endsAt'] as String),
    location: json['location'] as String,
    locationAr: json['locationAr'] as String?,
    isOfficial: json['isOfficial'] as bool,
    weightCategoriesOffered: ((json['weightCategoriesOffered'] as List?) ?? [])
        .map((e) => WeightCategory.fromJson(e as String))
        .toList(),
    level: json['level'] != null
        ? EventLevel.fromJson(json['level'] as String)
        : null,
    sanctioningBody: json['sanctioningBody'] as String?,
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'nameAr': nameAr,
    'hostingFederation': hostingFederation.name,
    'startsAt': startsAt.toIso8601String(),
    'endsAt': endsAt.toIso8601String(),
    'location': location,
    'locationAr': locationAr,
    'isOfficial': isOfficial,
    'weightCategoriesOffered': weightCategoriesOffered
        .map((w) => w.name)
        .toList(),
    'level': level?.name,
    'sanctioningBody': sanctioningBody,
  };
}

enum RegistrationStatus {
  registered,
  weighedIn,
  withdrawn,
  disqualified;

  String get labelKey => 'registration.$name';

  static RegistrationStatus fromJson(String raw) =>
      RegistrationStatus.values.firstWhere(
        (e) => e.name == raw,
        orElse: () => RegistrationStatus.registered,
      );
}

class TournamentRegistration {
  final EntityID id;
  final EntityID tournamentId;
  final EntityID athleteId;
  final WeightCategory weightCategory;
  final int? seedRank;
  final DateTime registeredAt;
  final RegistrationStatus status;

  // === per-event result ===
  final AgeGroup? ageDivisionEntered;
  final int? bracketSize;
  final int? finalPosition;
  final MedalType? medal;

  TournamentRegistration({
    EntityID? id,
    required this.tournamentId,
    required this.athleteId,
    required this.weightCategory,
    this.seedRank,
    DateTime? registeredAt,
    this.status = RegistrationStatus.registered,
    this.ageDivisionEntered,
    this.bracketSize,
    this.finalPosition,
    this.medal,
  }) : id = id ?? newEntityId(),
       registeredAt = registeredAt ?? DateTime.now();

  factory TournamentRegistration.fromJson(Map<String, dynamic> json) =>
      TournamentRegistration(
        id: json['id'] as String,
        tournamentId: json['tournamentID'] as String,
        athleteId: json['athleteID'] as String,
        weightCategory: WeightCategory.fromJson(
          json['weightCategory'] as String,
        ),
        seedRank: json['seedRank'] as int?,
        registeredAt: DateTime.parse(json['registeredAt'] as String),
        status: RegistrationStatus.fromJson(
          json['status'] as String? ?? 'registered',
        ),
        ageDivisionEntered: json['ageDivisionEntered'] != null
            ? AgeGroup.fromJson(json['ageDivisionEntered'] as String)
            : null,
        bracketSize: json['bracketSize'] as int?,
        finalPosition: json['finalPosition'] as int?,
        medal: json['medal'] != null
            ? MedalType.fromJson(json['medal'] as String)
            : null,
      );

  Map<String, dynamic> toJson() => {
    'id': id,
    'tournamentID': tournamentId,
    'athleteID': athleteId,
    'weightCategory': weightCategory.name,
    'seedRank': seedRank,
    'registeredAt': registeredAt.toIso8601String(),
    'status': status.name,
    'ageDivisionEntered': ageDivisionEntered?.name,
    'bracketSize': bracketSize,
    'finalPosition': finalPosition,
    'medal': medal?.name,
  };
}

class WeightCutEntry {
  final EntityID id;
  final EntityID registrationId;
  final DateTime recordedAt;
  final double currentKg;
  final double targetKg;
  final String? notes;

  const WeightCutEntry({
    required this.id,
    required this.registrationId,
    required this.recordedAt,
    required this.currentKg,
    required this.targetKg,
    this.notes,
  });

  factory WeightCutEntry.create({
    EntityID? id,
    required EntityID registrationId,
    required DateTime recordedAt,
    required double currentKg,
    required double targetKg,
    String? notes,
  }) => WeightCutEntry(
    id: id ?? newEntityId(),
    registrationId: registrationId,
    recordedAt: recordedAt,
    currentKg: currentKg,
    targetKg: targetKg,
    notes: notes,
  );

  double get deltaKg => currentKg - targetKg;

  int daysToCompetition(DateTime competitionDate) {
    return competitionDate.difference(recordedAt).inDays;
  }

  factory WeightCutEntry.fromJson(Map<String, dynamic> json) => WeightCutEntry(
    id: json['id'] as String,
    registrationId: json['registrationID'] as String,
    recordedAt: DateTime.parse(json['recordedAt'] as String),
    currentKg: (json['currentKg'] as num).toDouble(),
    targetKg: (json['targetKg'] as num).toDouble(),
    notes: json['notes'] as String?,
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'registrationID': registrationId,
    'recordedAt': recordedAt.toIso8601String(),
    'currentKg': currentKg,
    'targetKg': targetKg,
    'notes': notes,
  };
}

class Bracket {
  final EntityID id;
  final EntityID tournamentId;
  final WeightCategory weightCategory;
  final List<EntityID> seeds;
  final DateTime generatedAt;

  const Bracket({
    required this.id,
    required this.tournamentId,
    required this.weightCategory,
    required this.seeds,
    required this.generatedAt,
  });

  factory Bracket.create({
    EntityID? id,
    required EntityID tournamentId,
    required WeightCategory weightCategory,
    required List<EntityID> seeds,
    DateTime? generatedAt,
  }) => Bracket(
    id: id ?? newEntityId(),
    tournamentId: tournamentId,
    weightCategory: weightCategory,
    seeds: seeds,
    generatedAt: generatedAt ?? DateTime.now(),
  );

  factory Bracket.fromJson(Map<String, dynamic> json) => Bracket(
    id: json['id'] as String,
    tournamentId: json['tournamentID'] as String,
    weightCategory: WeightCategory.fromJson(json['weightCategory'] as String),
    seeds: ((json['seeds'] as List?) ?? []).map((e) => e as String).toList(),
    generatedAt: DateTime.parse(json['generatedAt'] as String),
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'tournamentID': tournamentId,
    'weightCategory': weightCategory.name,
    'seeds': seeds,
    'generatedAt': generatedAt.toIso8601String(),
  };
}

class BracketMatch {
  final EntityID id;
  final EntityID bracketId;
  final int round;
  final int position;
  final EntityID? athleteAId;
  final EntityID? athleteBId;
  final EntityID? winnerId;
  final EntityID? matchId;

  const BracketMatch({
    required this.id,
    required this.bracketId,
    required this.round,
    required this.position,
    this.athleteAId,
    this.athleteBId,
    this.winnerId,
    this.matchId,
  });

  factory BracketMatch.create({
    EntityID? id,
    required EntityID bracketId,
    required int round,
    required int position,
    EntityID? athleteAId,
    EntityID? athleteBId,
    EntityID? winnerId,
    EntityID? matchId,
  }) => BracketMatch(
    id: id ?? newEntityId(),
    bracketId: bracketId,
    round: round,
    position: position,
    athleteAId: athleteAId,
    athleteBId: athleteBId,
    winnerId: winnerId,
    matchId: matchId,
  );

  factory BracketMatch.fromJson(Map<String, dynamic> json) => BracketMatch(
    id: json['id'] as String,
    bracketId: json['bracketID'] as String,
    round: json['round'] as int,
    position: json['position'] as int,
    athleteAId: json['athleteAID'] as String?,
    athleteBId: json['athleteBID'] as String?,
    winnerId: json['winnerID'] as String?,
    matchId: json['matchID'] as String?,
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'bracketID': bracketId,
    'round': round,
    'position': position,
    'athleteAID': athleteAId,
    'athleteBID': athleteBId,
    'winnerID': winnerId,
    'matchID': matchId,
  };
}
